modal/cloud_bucket_mount.py

```diff
@@ -1,85 +1,105 @@
-
-
 @dataclass
 class _CloudBucketMount:
     """Mounts a cloud bucket to your container. Currently supports AWS S3 buckets.
 
-    S3 buckets are mounted using [AWS' S3 Mountpoint](https://github.com/awslabs/mountpoint-s3).
+    S3 buckets are mounted using [AWS S3 Mountpoint](https://github.com/awslabs/mountpoint-s3).
     S3 mounts are optimized for reading large files sequentially. It does not support every file operation; consult
     [the AWS S3 Mountpoint documentation](https://github.com/awslabs/mountpoint-s3/blob/main/doc/SEMANTICS.md) for more information.
 
-    **Usage**
+    **AWS S3 Usage**
 
     ```python
-    import modal
     import subprocess
 
     stub = modal.Stub()
+    secret = modal.Secret.from_dict({
+        "AWS_ACCESS_KEY_ID": "...",
+        "AWS_SECRET_ACCESS_KEY": "...",
+    })
+    @stub.function(
+        volumes={
+            "/my-mount": modal.CloudBucketMount(
+                bucket_name="s3-bucket-name",
+                secret=secret,
+                read_only=True
+            )
+        }
+    )
+    def f():
+        subprocess.run(["ls", "/my-mount"], check=True)
+    ```
+
+    **Cloudflare R2 Usage**
 
+    Cloudflare R2 is [S3-compatible](https://developers.cloudflare.com/r2/api/s3/api/) so its setup looks very similar to S3.
+    But additionally the `bucket_endpoint_url` argument must be passed.
+
+    ```python
+    import subprocess
+
+    stub = modal.Stub()
+    secret = modal.Secret.from_dict({
+        "AWS_ACCESS_KEY_ID": "...",
+        "AWS_SECRET_ACCESS_KEY": "...",
+    })
     @stub.function(
         volumes={
-            "/my-mount": modal.CloudBucketMount("s3-bucket-name", secret=modal.Secret.from_dict({
-                "AWS_ACCESS_KEY_ID": "...",
-                "AWS_SECRET_ACCESS_KEY": "...",
-            }), read_only=True)
+            "/my-mount": modal.CloudBucketMount(
+                bucket_name="my-r2-bucket",
+                bucket_endpoint_url="https://<ACCOUNT ID>.r2.cloudflarestorage.com",
+                secret=secret,
+                read_only=True
+            )
         }
     )
     def f():
-        subprocess.run(["ls", "/my-mount"])
+        subprocess.run(["ls", "/my-mount"], check=True)
     ```
     """
 
     bucket_name: str
+    # Endpoint URL is used to support Cloudflare R2.
+    bucket_endpoint_url: Optional[str]
 
-    # Credentials used to access a cloud bucket. When
-    # The given secret can contain AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.
-    # AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY can be omitted if the bucket is publicly accessible.
+    # Credentials used to access a cloud bucket.
+    # If the bucket is private, the secret **must** contain AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.
+    # If the bucket is publicly accessible, the secret is unnecessary and can be omitted.
     secret: Optional[_Secret] = None
 
     read_only: bool = False
     requester_pays: bool = False
-    bucket_type: Union[
```

Comment:
I think this is currently not needed, so will follow YAGNI and remove it. It's unnecessary to make the user pass this type explicitly when we can just infer it from `bucket_endpoint_url`.